### PR TITLE
Configurable Custom Components

### PR DESCRIPTION
**Pull request by loicteixeira**
Ref #9.

Inspired from `netlify-cms`, both for the registry and for exposing React functions.

It could use a bit more documentation on how to create an entity source, but I wonder if this should be a link to `springload/draftail` documentation (which does not exists FWIW) rather that on this project.

Also, previous discussion mentioned also publishing `wagtaildraftail` as a npm package for users who have their own build step and want to go deeper.
However, the only thing that felt missing was the `ModalSource` which is a wrapper around Wagtail's jQuery chooser modals.
Sure it would also be cool to have the registry available as a normal function, but it's available globally now so not essential.
In other words, I think it would be a nice to have, not a must have.

### CODE REVIEW

File: wagtaildraftail/client/wagtaildraftail.js [hunk 1]
```diff
@@ -54,4 +64,18 @@ const initDraftailEditor = (fieldName, options = {}) => {
...
+  registerDecorator: registerDecorator,
+  getDecorator: getDecorator,
+  registerSource: registerSource,
+  getSource: getSource,
+  registerStrategy: registerStrategy,
+  getStrategy: getStrategy,
+
+  // Expose basic React methods for basic needs
+  createClass: React.createClass,
+  createElement: React.createElement,
```
Comment (thibaudcolas) [inline wagtaildraftail/client/wagtaildraftail.js]: I don't think this is a good idea – users should be asked to use React directly to create their components (and they can do so without JSX if they wish to).

File: README.rst [hunk 2]
```diff
@@ -17,7 +17,7 @@ Check out `Awesome Wagtail <https://github.com/springload/awesome-wagtail>`_ for
 Installation
 ------------
 
-Grab the package from pip with ``pip install wagtaildraftail``, then add ``wagtaildraftail`` as an app in your Django settings.
+Grab the package from pip with ``pip install wagtaildraftail``, then add ``wagtaildraftail`` (before any other app which will try to register an entity) as an app in your Django settings.
```
Comment (thibaudcolas) [inline README.rst]: This strikes me as confusing language for the most common use-case where people do not want to make their own components. Could we move this to the dedicated documentation section?

File: README.rst [hunk 3]
```diff
@@ -218,7 +218,79 @@ TODO
...
 
-TODO
+An entity basically needs 4 elements:
+-  a page ``decorator`` to render the entity on the page (implemented in Python).
+-  an editor ``decorator`` to render the entity in the editor (implemented in JS).
+-  an editor ``source`` to select an entity (implemented in JS), e.g. a modal to select a page for a link.
+-  an editor ``strategy`` to find the entity when the editor is loaded (implemented in JS), which is optional, the default one works fine in most cases.
+
+A ``page decorator`` is a simple Python class with a ``render`` method accepting a single argument, the ``props`` of the element. The ``render`` method can either use the ``DOM.create_element`` from ``draftjs_exporter`` (which mimic ``React.createElement``) or render some html directly. It is registered in ``DRAFT_EXPORTER_ENTITY_DECORATORS`` in the `config file <#configuration>`_.
+
```
Comment (thibaudcolas) [inline README.rst]: I think I would rather have this documentation pointing at https://github.com/springload/draftjs_exporter, with the least overlap in content, so all of this content is easier to maintain over time. For this section of the documentation, this would mean not having the code sample and then for the text going along those lines:

> Decorators define how the content needs to be displayed on the site's pages, as well as within the editor.
>
> - For the pages, they are defined in Python with `draftjs_exporter`. Refer to the dedicated documentation on [the draftjs_exporter README](https://github.com/springload/draftjs_exporter#custom-components).
> - For the editor, they are defined in JS/React with `draftail` Refer to the dedicated documentation on [the Draftail README](https://github.com/springload/draftail).

We would still need to describe the `wagtaildraftail` registration / config here, since this is what this particular project "owns".

NB: I do think a big walkthrough would be valuable, but it would be more appropriate in a separate (and potentially one-off) tutorial or blog post, rather than in the README.

File: README.rst [hunk 4]
```diff
@@ -218,7 +218,79 @@ TODO
...
+      entityKey: React.PropTypes.string.isRequired,
+      children: React.PropTypes.node.isRequired,
+    };
+
+    // Or `export default Link;` and register later.
+    window.wagtaildraftail.registerDecorator('Link', Link);
+
+    /* More examples at https://github.com/springload/wagtaildraftail/tree/master/wagtaildraftail/client/decorators */
+
+An ``editor source`` is usually more complex and will usually show a modal for the user to select an object or some options.
```
Comment (thibaudcolas) [inline README.rst]: I'll detail this further either here or in Draftail so it's clearer what's going on.

File: tests/testapp/home/decorators.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+                page = Page.objects.get(id=page_id)
+                href = page.url
+            except Page.DoesNotExist:
+                href = props.get('url', MISSING_RESOURCE_URL)
+        else:
+            href = props.get('url', MISSING_RESOURCE_URL)
+
+        anchor_properties = {
+            'className': 'button',
+            'onclick': "window.location.href='{}'".format(href),
```
Comment (thibaudcolas) [inline tests/testapp/home/decorators.py]: I know it's just example / test code but we should follow best practice nonetheless - do not use `onclick`.

File: tests/testapp/testapp/grains/rich_text.py [hunk 6]
```diff
@@ -47,6 +47,13 @@
     'source': 'DocumentSource',
     'decorator': 'Document',
 }
+DRAFT_ENTITY_TYPE_BUTTON = {
+    'label': 'Button',
+    'type': 'BUTTON',
```
Comment (thibaudcolas) [inline tests/testapp/testapp/grains/rich_text.py]: I see this string/token is duplicated. This is the code people will likely look at / we will showcase in our examples, do we want to encourage people to set up their own constants for those types, or simply use strings? What's the most Pythonic?

File: wagtaildraftail/client/registry.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+const _registry = {
```
Comment (thibaudcolas) [inline wagtaildraftail/client/registry.js]: No underscore dangle in JS.

File: wagtaildraftail/client/registry.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+const registerDecorator = (name, decorator) => {
+  _registry.decorators[name] = decorator;
+  console.log('decorator: ' + name);
+};
+
+const getDecorator = (name) => {
+  return _registry.decorators[name];
+};
```
Comment (thibaudcolas) [inline wagtaildraftail/client/registry.js]: Should be an arrow function.

Edit: soz, `inline` arrow function.

File: wagtaildraftail/client/registry.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
+const _registry = {
+  decorators: {},
+  sources: {},
+  strategies: {},
+};
+
+const registerDecorator = (name, decorator) => {
+  _registry.decorators[name] = decorator;
+  console.log('decorator: ' + name);
+};
```
Comment (thibaudcolas) [inline wagtaildraftail/client/registry.js]: I think the signature of those `register` functions could be made simpler by taking advantage of object shorthands. New signature:

```js
const registerDecorators = (decorators) => {
  Object.assign(_registry.decorators, decorators);
};

window.wagtaildraftail.registerDecorators({ ButtonDecorator });
```

This has the advantage that you don't have to repeat the decorator name and its value. The other advantage is that it promotes registering all of your decorators / strategies / etc at once – we want to encourage people to separate the definition of those components from the side effect of registering them. This would look like:

```js
// In the JS entry bundle.
import ButtonDecorator from './ButtonDecorator';
import LocationDecorator from './LocationDecorator';

window.wagtaildraftail.registerDecorators({
   ButtonDecorator,
   LocationDecorator,
});
```

File: wagtaildraftail/client/registry.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+export {
+  registerDecorator,
+  getDecorator,
+  registerSource,
+  getSource,
+  registerStrategy,
+  getStrategy,
+}
```
Comment (thibaudcolas) [inline wagtaildraftail/client/registry.js]: This really looks like one object rather than a bunch of named exports – could we `export default` a single object instead? Then in the respective import, `Object.assign` its methods to the `wagtailDraftail` export.

File: wagtaildraftail/client/wagtaildraftail.js [hunk 11]
```diff
@@ -6,11 +6,20 @@ import DraftailEditor from 'draftail';
...
+import sources from './sources';
+import {registerDecorator, getDecorator, registerSource, getSource, registerStrategy, getStrategy} from './registry';
+
+// Register Decorators, Sources and Strategies
+Object.keys(decorators).forEach(function(key) {
+  registerDecorator(key, decorators[key]);
+});
+Object.keys(sources).forEach(function(key) {
+  registerSource(key, sources[key]);
+});
```
Comment (thibaudcolas) [inline wagtaildraftail/client/wagtaildraftail.js]: Those two constructs would be simplified as well by using the `registerDecorators` proposal above.

File: wagtaildraftail/client/wagtaildraftail.js [hunk 12]
```diff
@@ -35,9 +45,9 @@ const initDraftailEditor = (fieldName, options = {}) => {
     // eslint-disable-next-line no-param-reassign
     options.entityTypes = options.entityTypes.map(entity => Object.assign(entity, {
       // TODO: Rename keys accordingly once changed in draftail https://github.com/springload/draftail/issues/49
-      control: sources[entity.source],
-      strategy: getEntityStrategy(entity.type),
-      component: decorators[entity.decorator],
+      control: getSource(entity.source),
+      strategy: getStrategy(entity.type) || getDefaultStrategy(entity.type),
+      component: getDecorator(entity.decorator),
```
Comment (thibaudcolas) [inline wagtaildraftail/client/wagtaildraftail.js]: I think the `wagtaildraftail` object should be defined above, so this code can use the same `wagtaildraftail.getSource` and so on as the userland code.

Comment (thibaudcolas) [reply wagtaildraftail/client/wagtaildraftail.js]: Additional note (that can wait for another PR): do we want to provide a way to define fallbacks for all of those things, like we do for the exporter?

File: wagtaildraftail/client/wagtaildraftail.js [hunk 13]
```diff
@@ -54,4 +64,18 @@ const initDraftailEditor = (fieldName, options = {}) => {
 
 window.initDraftailEditor = initDraftailEditor;
 
+window.wagtaildraftail = {
```
Comment (thibaudcolas) [inline wagtaildraftail/client/wagtaildraftail.js]: Should this be `wagtaildraftail` or `wagtailDraftail`? The latter comes naturally to me from JS land.

File: tests/testapp/home/static/button_entity.js [hunk 14]
```diff
@@ -0,0 +1,6 @@
+const ButtonDecorator = ({ entityKey, children }) => {
+  const attrs = {'data-tooltip': entityKey, className: 'RichEditor-button'};
+  return window.wagtaildraftail.createElement('span', attrs, children);
+};
```
Comment (thibaudcolas) [inline tests/testapp/home/static/button_entity.js]: Honestly I'm not sure I want to encourage people to do this without a build step. It's fine for this PR but I might add one later on.

**Review by thibaudcolas — DISMISSED** [review]
Thank you for taking care of this! Here is some feedback before I head to bed. I'll get back to all of this tomorrow.

File: tests/testapp/home/decorators.py (hunk 5 above)
Comment (loicteixeira) [reply tests/testapp/home/decorators.py]: I was struggling to come up with an element which would reuse an existing `source` and still do something interesting, but I'll just convert this to an anchor with a regular href, explaining this is just for demo purposes.

File: tests/testapp/home/static/button_entity.js (hunk 14 above)
Comment (loicteixeira) [reply tests/testapp/home/static/button_entity.js]: I thought we wanted to allow an easy path, but maybe that's too easy 😉

To remove this though, I'll need help with the `webpack` config, so it has 2 entry points (the admin and the test sites) and therefore 2 exit points.

File: wagtaildraftail/client/registry.js (hunk 7 above)
Comment (loicteixeira) [reply wagtaildraftail/client/registry.js]: This is why you shouldn't let Python developer do JS 😮

File: tests/testapp/testapp/grains/rich_text.py (hunk 6 above)
Comment (loicteixeira) [reply tests/testapp/testapp/grains/rich_text.py]: Pythonic or not, using strings here can indeed be considered bad practice so I'll change that.

File: wagtaildraftail/client/wagtaildraftail.js (hunk 12 above)
Comment (loicteixeira) [reply wagtaildraftail/client/wagtaildraftail.js]: I disagree with using `window.wagtaildraftail.getSource` in the project's code. Sure it will make it easier for a person reading the source code that it's using the exact same function that the one they use in their custom components, but it would make the whole thing less readable and very verbose for a very limited number of people (who go through the source code rather than the documentation).

No objection to fallbacks.

File: wagtaildraftail/client/wagtaildraftail.js [hunk 15]
```diff
@@ -54,4 +64,18 @@ const initDraftailEditor = (fieldName, options = {}) => {
 
 window.initDraftailEditor = initDraftailEditor;
```
Comment (loicteixeira) [inline wagtaildraftail/client/wagtaildraftail.js]: @thibaudcolas: Do we want this to be `window.wagtailDraftail.initEditor` now that we have a namespace for other utils?

File: wagtaildraftail/client/wagtaildraftail.js (hunk 1 above)
Comment (loicteixeira) [reply wagtaildraftail/client/wagtaildraftail.js]: See my reply to the similar comment below on `tests/testapp/home/static/button_entity.js`.

File: wagtaildraftail/client/wagtaildraftail.js (hunk 15 above)
Comment (thibaudcolas) [reply wagtaildraftail/client/wagtaildraftail.js]: Yup, makes sense to me!

File: tests/testapp/home/static/button_entity.js (hunk 14 above)
Comment (thibaudcolas) [reply tests/testapp/home/static/button_entity.js]: I think it's too brittle. To write even this small amount of JS/React without a compilation step, this means:

- Loading React on the page in a global variable.
- Writing React code without JSX.
- Writing JS code without module bundling.
- Writing JS code without transpiling.

In 2017 all of those things are bad practice IMHO. For example, this `ButtonDecorator` uses ES6 object destructuring and arrow functions so it won't work in IE11.

I'll have a think about how best to present a reusable API to create decorators and sources with the least amount of JS, for the purpose of this PR I'm happy to go ahead with this 👌

Comment (loicteixeira) [reply tests/testapp/home/static/button_entity.js]: Agreed.
